File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/KeyPointCenter/KeyPoint/LowPassFilter.swift
//
//  LowPassFilter.swift
//  MetaScripts
//
//  Created by ZhaoTianyu on 2024/7/10.
//

import Foundation

// LowPassFilter类
final class LowPassFilter {
    // Fields
    private var effectiveCount = 0
    private let FactorialCount = 10
    private var NOrderLPF: Int
    private var prevPos3D: [Vector3]
    private var Smooth: Float

    // Methods
    init() {
        self.prevPos3D = [Vector3](repeating: Vector3(x: 0, y: 0, z: 0), count: 10)
        self.NOrderLPF = 7
        self.Smooth = 0.9
    }

    init(order: Int, smooth: Float) {
        self.prevPos3D = [Vector3](repeating: Vector3(x: 0, y: 0, z: 0), count: 10)
        self.NOrderLPF = min(order, 10)
        self.Smooth = smooth
    }

    func correctAndPredict(kp: Vector3) -> Vector3 {
        self.prevPos3D[0] = kp
        for i in 1..<self.NOrderLPF {
            self.prevPos3D[i] = (self.prevPos3D[i] * (1 - self.Smooth)) + (self.prevPos3D[i - 1] * self.Smooth)
        }
        self.prevPos3D[0] = (self.prevPos3D[0] * (1 - self.Smooth)) + (self.prevPos3D[self.NOrderLPF - 1] * self.Smooth)
        if self.effectiveCount < 10 {
            self.effectiveCount += 1
            return kp
        }
        return self.prevPos3D[0]
    }
}

File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/UnityDEBUG/ForDebug/Assets/Scripts/CharacterDriven/LowPassFilter.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
public sealed class LowPassFilter
{
    // Fields
    private int effectiveCount;
    private const int FactorialCount = 10;
    private int NOrderLPF;
    private Vector3[] prevPos3D;
    private float Smooth;

    // Methods
    public LowPassFilter()
    {
        this.prevPos3D = new Vector3[10];
        this.NOrderLPF = 7;
        this.Smooth = 0.9f;
    }

    public LowPassFilter(int order, float smooth)
    {
        this.prevPos3D = new Vector3[10];
        this.NOrderLPF = 7;
        this.Smooth = 0.9f;
        this.NOrderLPF = Mathf.Min(order, 10);
        this.Smooth = smooth;
    }

    public Vector3 CorrectAndPredict(Vector3 kp)
    {
        this.prevPos3D[0] = kp;
        for (int i = 1; i < this.NOrderLPF; i++)
        {
            this.prevPos3D[i] = (Vector3)((this.prevPos3D[i] * (1f - this.Smooth)) + (this.prevPos3D[i - 1] * this.Smooth));
        }
        this.prevPos3D[0] = (Vector3)((this.prevPos3D[0] * (1f - this.Smooth)) + (this.prevPos3D[this.NOrderLPF - 1] * this.Smooth));
        if (this.effectiveCount < 10)
        {
            this.effectiveCount++;
            return kp;
        }
        return this.prevPos3D[0];
    }
}


